src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.T.cs

```diff
@@ -0,0 +1,114 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// </summary>
+    /// <remarks>
+    /// <para>This type corresponds to <see cref="GCHandleType.Normal"/>.</para>
+    /// <para>
+    /// This type is unsafe if used incorrectly. Incorrect usage like mismanagement
+    /// of lifetime, providing invalid handle value or concurrent disposal can result
+    /// into hard to diagnose crashes or data corruptions.
+    /// </para>
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct GCHandle<T> : IEquatable<GCHandle<T>>, IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public GCHandle(T target)
+        {
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Normal);
```

Comment:
Since we rely on IntPtr.Zero elsewhere in the implementation having special meaning, worth asserting here that _handle != IntPtr.Zero?

Reply:
Should the assert go to `InternalAlloc` since there are more types depending on this?